endorsedLibraries/kotlinx.cli/src/main/kotlin/kotlinx/cli/Arguments.kt

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package kotlinx.cli
+
+import kotlin.reflect.KProperty
+
+internal data class CLIEntityWrapper(var entity: CLIEntity<*>? = null)
+
+/**
+ * Command line entity.
+ *
+ * @property owner parser which owns current entity.
+ */
+abstract class CLIEntity<TResult> internal constructor(internal val owner: CLIEntityWrapper) {
+    /**
+     * CLI element value.
+     */
+    internal lateinit var cliElement: ArgumentValueDelegate<TResult>
+    /**
+     * Wrapper  for element - read only property.
+     * Needed to close set of variable [cliElement].
+     */
+    val delegate: ArgumentValueDelegate<TResult>
+        get() = cliElement
+
+    /**
+     * Value of entity.
+     */
+    var value: TResult
+        get() = cliElement.value
+        set(value) = delegate.setValue(null, ::value, value)
+
+    /**
+     * Origin of argument value.
+     */
+    val valueOrigin: ArgParser.ValueOrigin
+        get() = (delegate as ParsingValue<*, *>).valueOrigin
+
+    operator fun provideDelegate(thisRef: Any?, prop: KProperty<*>): ArgumentValueDelegate<TResult> {
+        (cliElement as ParsingValue<*, *>).provideName(prop.name)
+        return cliElement
+    }
+}
+
+/**
+ * Argument instance.
+ */
+abstract class Argument<TResult> internal constructor(owner: CLIEntityWrapper): CLIEntity<TResult>(owner)
+
+/**
+ * Common single argument instance.
+ */
+abstract class AbstractSingleArgument<T: Any, TResult> internal constructor(owner: CLIEntityWrapper): Argument<TResult>(owner) {
+    /**
+     * Check descriptor for this kind of argument.
+     */
+    internal fun checkDescriptor(descriptor: ArgDescriptor<*, *>) {
+        if (descriptor.number == null || descriptor.number > 1) {
+            error("Argument with single value can't be initialized with descriptor for multiple values.")
```

Comment:
Could this be a user error or it's just our internal assertion?

Reply:
That's internal assertion, I don't know if we want to make library real MPP. If so JS has now assert method

Reply:
We can introduce `fun failAssertion(message) = throw AssertionError(message)` and use it just so the intent is clear.

Reply:
Fixed